articles/validating-xml-using-dtd/index.md

```diff
@@ -0,0 +1,250 @@
+<customer>
+  <firstname>John</firstname>
+  <lastname>Doe</lastname>
+  <companyname>Section</companyname>
+  <email>johndoe@section.io</email>
+  <message>Welcome message</message>
+</customer>
+```
+
+We have built our first XML document. Now, it can be saved locally with a `.xml` file extension.
+
+#### Building DTD validator
+Before we start building, let us understand a few keywords, that would help us understand better.
+
+##### Elements
+Elements are the building blocks of DTD. In HTML, you would have noticed elements like "p" and "table" for holding our data. But, in XML, the elements refer to "firstname" and "email".
+
+Example in HTML:
+
+```html
+<h1>Heading</h1>
+<p>Hello world!</p>
+```
+
+Example in XML:
+
+```xml
+<companyname>Section</companyname>
+<email>johndoe@section.io</email>
+```
+
+##### Attributes
+Attributes provide extra information about the elements. It is always placed inside the tag of elements and is in the form of a name-value pair.
+
+Example in HTML:
+
+```html
+<img src='helloworld.png' />
+```
+
+Example in XML:
+
+```xml
+<message id='1'>Hello World</message>
+```
+
+##### PCDATA and CDATA
+PCDATA is the Parsed Character Data. It is a text which will be parsed by the DTD parser. The tags inside the document are parsed to retrieve the values inside each of the tags, for validation.
+
+Similarly, CDATA is Character Data. It is a text which will not be parsed by the DTD parser. The elements will not be parsed, and the values cannot be retrieved.
+
+#### Validating using DTD
+Generally, there are two types of DTD validations: Internal validation and External validation.
+
+For internal validations, the whole DTD will be written in the same file as that of the XML file, which can be used for validation.
+
+Similarly, the external validation will validate the XML based on the DTD written in a separate file with the `.dtd` extension. For linking it with the XML file, we use
+
+```xml
+<!DOCTYPE customer SYSTEM "validation.dtd">
+```
+
+Having understood various components of DTD, let's start building a DTD validator.
+
+In DTD, the XML elements are declared as
+
+```xml
+<!ELEMENT element-name (element-content)>
+```
+
+For our example, the outermost tag `<customer />` holds all the input fields together. So, in DTD, we must specify the tag as `<customer />` containing fields with tags like `<firstname />`, `<lastname />`, `<companyname />`, `<email />` and `<message />`. According to the DTD syntax, we must specify that `<customer />` tag as the parent element holding other child tags.
+
+```xml
+<!ELEMENT customer (firstname, lastname, companyname, email, message)>
+```
+
+After specifying the tags present inside the outermost tag. Now, we have to define the type of value that can be parsed for each tag. Here, all the input fields can contain data as text. According to DTD syntax, we specify it as `#PCDATA`, under each element. Since it is required to parse for validating if the values are textual data or not.
```

Comment:
> Since it is required to parse for validating if the values are textual data or not.

You can change this to sentence in a way that flows better like:

**This is done since these values should be parsed and validated.**